### ISSUE TITLE
Making abstract class package private makes generated class' copyOf() method package private.

### ISSUE THREAD

**Issue by matt-isaman** [immutables/immutables#687]
I am attempting to use Immutables (version 2.5.5) to define an immutable set of domain objects for a public interface. I would like to hide as much as possible from the users about the fact that I am generating these classes. As such I want to mark the abstract classes package private and the generated classes public. I noticed that when I did this the `copyOf()` method on the generated class also became package private. I could be fine not having this convenience method, but if possible I would like it there.

This class:
```java
@Immutable
@Style(typeAbstract = "*Defn", typeImmutable = "*")
public abstract class FooDefn { ... }
```
Generates:
```java
public final class Foo extends FooDefn {
  ...
  public static Foo copyOf(FooDefn instance) { ... }
  ...
  public static final class Builder {
    ...
    public final Builder from(FooDefn instance) { ... }
    ...
  }
}
```
I didn't like this because it exposes the `FooDefn` class in the method signatures and as I mentioned I want to hide that from users.

After poking around some I changed things to the following:
```java
@Immutable
@Style(typeAbstract = "*Defn", typeImmutable = "*", visibility = PUBLIC)
abstract class FooDefn { ... }
```
Which generated:
```java
public final class Foo extends FooDefn {
  ...
  static Foo copyOf(FooDefn instance) { ... }
  ...
  public static final class Builder {
    ...
    // the below maintains the method but hides the abstract class, nice!
    public final Builder from(Foo instance) { ... }
    final Builder from(FooDefn instance) { ... }
    ...
  }
}
```

It seems to me this generated class should have an extra method with the signature:
```java
  public static Foo copyOf(Foo instance) { ... }
```
This would match what happened with the `Builder#from()` method and would allow me to use all of the functionality provided by Immutables without exposing the underlying abstract classes. After searching around it seems like this might be a regression of a change mentioned at the bottom of #236.

Another option is to make both the `copyOf()` and `Builder#from()` methods just take the generated class type in all cases. Why would we want the abstract classes referenced from the generated class? Other than comments and the extends clause of course. There is perhaps a reason to not do this that I am unaware of but I would prefer these methods take the generated class.

**Comment by matt-isaman** [comment]
Now that I look at the generated code I would also prefer that the method comments in the generated class never referenced the abstract class.

This for example:
```java
    /**
     * Initializes the value for the {@link FooDefn#getFoo() foo} attribute.
     * @param foo The value for foo 
     * @return {@code this} builder for use in a chained invocation
     */
    public final Builder foo(int foo) { ... }
```

In my opinion should be:
```java
    /**
     * Initializes the value for the {@link Foo#getFoo() foo} attribute.
     * @param foo The value for foo 
     * @return {@code this} builder for use in a chained invocation
     */
    public final Builder foo(int foo) { ... }
```

This also applies to the comments on the following methods:
```java
    public final Foo withFoo(int value) { ... }
    public static Foo copyOf(Foo instance) { ... } // assuming the initial request in this issue is fulfilled
```

There are perhaps other examples not generated by the simple class I have used for this issue.

**Comment by elucash** [comment]
(sorry, I'm getting to only now)
```
public static Foo copyOf(Foo instance) { ... } 
```
this method is not generated because it always returns the same object and is useless. While package private `static Foo copyOf(FooDefn instance) { ... }` can handle other subclasses of `FooDefn` if the calling code happens to be in the same package (like some test objects). Builder's `Builder from(Foo foo)` is useful because it fills builder with the data from an instance of Foo, so it's not a no-op method.

about usages in Javadoc it's definitely a problem, to be fixed in 2.6. Thank you!
